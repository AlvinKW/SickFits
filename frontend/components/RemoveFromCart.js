import React, { Component } from 'react';
import gql from 'graphql-tag';
import styled from 'styled-components';
import { Mutation } from 'react-apollo';
import PropTypes from 'prop-types';

import { CURRENT_USER_QUERY } from './User';

const REMOVE_FROM_CART_MUTATION = gql`
	mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
		removeFromCart(id: $id) {
			id
		}
	}
`;

const StyledButton = styled.button`
	border: none;
	font-size: 3rem;
	background: none;

	&:hover {
		color: ${props => props.theme.red};
		cursor: pointer;
	}
`;

class RemoveFromCart extends Component {
	render() {
		return (
			<Mutation
				mutation={REMOVE_FROM_CART_MUTATION}
				variables={{ id: this.props.id }}
			>
				{(removeFromCart, { loading }) => (
					<StyledButton
						title="Delete Item"
						disabled={loading}
						onClick={() => {
							removeFromCart().catch(error => alert(error.message));
						}}
					>
						&times;
					</StyledButton>
				)}
			</Mutation>
		);
	}
}

RemoveFromCart.propTypes = {
	id: PropTypes.string.isRequired,
};

export default RemoveFromCart;
