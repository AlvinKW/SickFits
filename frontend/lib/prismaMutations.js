import gql from 'graphql-tag';

export const ADD_TO_CART_MUTATION = gql`
	mutation ADD_TO_CART_MUTATION($id: ID!) {
		addToCart(id: $id) {
			id
			quantity
		}
	}
`;

export const CREATE_ITEM_MUTATION = gql`
	mutation CREATE_ITEM_MUTATION (
		$title: String!
		$description: String!
		$image: String
		$largeImage: String
		$price: Int!
	) {
		createItem(
			title: $title
			description: $description
			image: $image
			largeImage: $largeImage
			price: $price
		) {
			id
		}
	}
`;

export const CREATE_ORDER_MUTATION = gql`
	mutation CREATE_ORDER_MUTATION($token: String!) {
		createOrder(token: $token) {
			id
			total
			charge
			items {
				id
				title
			}
		}
	}
`;

export const DELETE_ITEM_MUTATION = gql`
	mutation DELETE_ITEM_MUTATION($id: ID!) {
		deleteItem(id: $id) {
			id
		}
	}
`;

export const REMOVE_FROM_CART_MUTATION = gql`
	mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
		removeFromCart(id: $id) {
			id
		}
	}
`;

export const REQUEST_RESET_MUTATION = gql`
	mutation REQUEST_RESET_MUTATION($email: String!) {
		requestReset(email: $email) {
			message
		}
	}
`;

export const RESET_PASSWORD_MUTATION = gql`
	mutation RESET_PASSWORD_MUTATION($resetToken: String!, $password: String!, $confirmPassword: String!) {
		resetPassword(resetToken: $resetToken, password: $password, confirmPassword: $confirmPassword) {
			id
			name
			email
		}
	}
`;

export const SIGN_IN_MUTATION = gql`
	mutation SIGNIN_MUTATION($email: String!, $password: String!) {
		signIn(email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

export const SIGN_OUT_MUTATION = gql`
	mutation SIGN_OUT_MUTATION {
		signOut {
			message
		}
	}
`;

export const SIGN_UP_MUTATION = gql`
	mutation SIGNUP_MUTATION($name: String!, $email: String!, $password: String!) {
		signUp(name: $name, email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

export const TOGGLE_CART_MUTATION = gql`
	mutation TOGGLE_CART_MUTATION {
		toggleCart @client
	}
`;

export const UPDATE_ITEM_MUTATION = gql`
	mutation UPDATE_ITEM_MUTATION (
		$id: ID!
		$title: String
		$description: String
		$price: Int
	) {
		updateItem(
			id: $id
			title: $title
			description: $description
			price: $price
		) {
			id
			title
			description
			price
		}
	}
`;

export const UPDATE_PERMISSIONS_MUTATION = gql`
	mutation UPDATE_PERMISSIONS_MUTATION($permissions: [Permission], $userID: ID!) {
		updatePermissions(permissions: $permissions, userID: $userID) {
			id
			name
			email
			permissions
		}
	}
`;
