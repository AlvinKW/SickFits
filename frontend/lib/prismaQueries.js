import gql from 'graphql-tag';

import { perPage } from '../config';

export const ALL_ITEMS_QUERY = gql`
	query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = ${perPage}) {
		items(orderBy: createdAt_DESC, skip: $skip, first: $first) {
			id
			title
			description
			image
			largeImage
			price
		}
	}
`;

export const ALL_USERS_QUERY = gql`
	query ALL_USERS_QUERY {
		users {
			id
			name
			email
			permissions
		}
	}
`;

export const CURRENT_USER_QUERY = gql`
	query CURRENT_USER_QUERY {
		me {
			id
			name
			email
			permissions
			orders {
				id
			}
			cart {
				id
				quantity
				item {
					id
					title
					description
					image
					price
				}
			}
		}
	}
`;

export const LOCAL_STATE_QUERY = gql`
	query LOCAL_STATE_QUERY {
		cartOpen @client
	}
`;

export const PAGINATION_QUERY = gql`
	query PAGINATION_QUERY {
		itemsConnection {
			aggregate {
				count
			}
		}
	}
`;

export const SEARCH_ITEMS_QUERY = gql`
	query SEARCH_ITEMS_QUERY($searchTerm: String!) {
		items(where: { OR: [{ title_contains: $searchTerm }, { description_contains: $searchTerm }] }) {
			id
			title
			image
		}
	}
`;

export const SINGLE_ITEM_QUERY = gql`
	query SINGLE_ITEM_QUERY($id: ID!) {
		item(where: { id: $id }) {
			id
			title
			description
			largeImage
			price
		}
	}
`;

export const SINGLE_ORDER_QUERY = gql`
	query SINGLE_ORDER_QUERY($id: ID!) {
		order(id: $id) {
			id
			total
			charge
			createdAt
			items {
				id
				title
				description
				image
				price
				quantity
			}
			user {
				id
			}
		}
	}
`;

export const USER_ORDERS_QUERY = gql`
	query USER_ORDERS_QUERY {
		orders(orderBy: createdAt_DESC) {
			id
			total
			createdAt
			items {
				id
				title
				description
				image
				price
				quantity
			}
		}
	}
`;
