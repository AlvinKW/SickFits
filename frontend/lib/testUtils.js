import casual from 'casual';

casual.seed(777);

const mockItem = () => ({
	__typename: 'Item',
	id: 'abc123',
	price: 5000,
	user: null,
	image: 'dog-small.jpg',
	title: 'Dogs are the best!',
	description: 'Dogs',
	largeImage: 'dog.jpg',
});

const mockUser = () => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: [],
});

const mockOrderItem = () => ({
	__typename: 'OrderItem',
	id: casual.uuid,
	image: `${casual.word}.jpg`,
	title: casual.words(),
	price: 4234,
	quantity: 1,
	description: casual.words(),
});

const mockOrder = () => ({
	__typename: 'Order',
	id: 'ord123',
	charge: 'ch_123',
	total: 40000,
	items: [mockOrderItem(), mockOrderItem()],
	createdAt: '2018-04 - 06T19: 24: 16.000Z',
	user: mockUser(),
});

const mockCartItem = overrides => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 3,
	item: mockItem(),
	user: mockUser(),
	...overrides,
});

class MockLocalStorage {
	constructor() {
		this.store = {};
	}

	clear() {
		this.store = {};
	}

	getItem(key) {
		return this.store[key] || null;
	}

	setItem(key, value) {
		this.store[key] = value.toString();
	}

	removeItem(key) {
		delete this.store[key];
	}
}

export {
	MockLocalStorage,
	mockItem,
	mockUser,
	mockCartItem,
	mockOrder,
	mockOrderItem,
};
