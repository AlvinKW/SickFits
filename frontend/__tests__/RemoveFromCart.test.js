import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import { ApolloConsumer } from 'react-apollo';

import { mockUser, mockCartItem } from '../lib/testUtils';
import { CURRENT_USER_QUERY } from '../lib/prismaQueries';
import { REMOVE_FROM_CART_MUTATION } from '../lib/prismaMutations';
import RemoveFromCart from '../components/RemoveFromCart';

const mocks = [
	{
		request: {
			query: CURRENT_USER_QUERY,
		},
		result: {
			data: {
				me: {
					...mockUser(),
					cart: [mockCartItem({ id: 'ABC123' })],
				},
			},
		},
	},
	{
		request: {
			query: REMOVE_FROM_CART_MUTATION,
			variables: { id: 'ABC123' },
		},
		result: {
			data: {
				removeFromCart: {
					__typename: 'CartItem',
					id: 'ABC123',
				},
			},
		},
	},
];

describe('<RemoveFromCart />', () => {
	it('renders and matches the snapshot', () => {
		const wrapper = mount(
			<MockedProvider>
				<RemoveFromCart id="ABC123" />
			</MockedProvider>
		);

		expect(toJSON(wrapper.find('button'))).toMatchSnapshot();
	});

	it('removes the item from the cart', async () => {
		let apolloClient = null;
		const wrapper = mount(
			<MockedProvider mocks={mocks}>
				<ApolloConsumer>
					{client => {
						apolloClient = client;
						return <RemoveFromCart id="ABC123" />;
					}}
				</ApolloConsumer>
			</MockedProvider>
		);

		const response = await apolloClient.query({ query: CURRENT_USER_QUERY });
		expect(response.data.me.cart).toHaveLength(1);
		expect(response.data.me.cart[0].item.price).toBe(5000);

		wrapper.find('button').simulate('click');
		await wait();
		const response2 = await apolloClient.query({ query: CURRENT_USER_QUERY });
		expect(response2.data.me.cart).toHaveLength(0);
	});
});
