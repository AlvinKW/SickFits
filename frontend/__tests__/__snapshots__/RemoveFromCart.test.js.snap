// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RemoveFromCart /> renders and matches the snapshot 1`] = `
<button
  className="RemoveFromCart__StyledButton-emvtd6-0 flIdSz"
  disabled={false}
  onClick={[Function]}
  title="Delete Item"
>
  Ã—
</button>
`;
